File: mockserver-client-php/src/LoadThreshold.php
<?php

declare(strict_types=1);

namespace MockServer;

/**
 * An in-run pass/fail threshold for a {@see LoadScenario}: a per-run metric
 * compared against a value. All thresholds must hold for the run verdict to be
 * PASS (logical AND); any breach makes the verdict FAIL. Evaluated from the
 * run's own latency histogram and counters, not the global SLO sample store.
 *
 * Use the static factory {@see LoadThreshold::of()}; it emits a camelCase array
 * with the {@code metric}, {@code comparator} and {@code threshold} fields.
 *
 * @example
 *   LoadThreshold::of('LATENCY_P95', 'LESS_THAN', 250);
 *   LoadThreshold::of('ERROR_RATE', 'LESS_THAN_OR_EQUAL', 0.01);
 */
class LoadThreshold implements \JsonSerializable
{
    private string $metric;
    private string $comparator;
    private float $threshold;

    private function __construct(string $metric, string $comparator, float $threshold)
    {
        $this->metric = strtoupper($metric);
        $this->comparator = strtoupper($comparator);
        $this->threshold = $threshold;
    }

    /**
     * Build a threshold from a per-run metric, a comparator and a value.
     *
     * @param string $metric One of LATENCY_P50, LATENCY_P95, LATENCY_P99,
     *        LATENCY_P999, ERROR_RATE, THROUGHPUT_RPS.
     * @param string $comparator One of LESS_THAN, LESS_THAN_OR_EQUAL,
     *        GREATER_THAN, GREATER_THAN_OR_EQUAL.
     * @param float $threshold The threshold value (milliseconds for latency
     *        metrics, a 0.0-1.0 fraction for ERROR_RATE, requests/second for
     *        THROUGHPUT_RPS).
     */
    public static function of(string $metric, string $comparator, float $threshold): self
    {
        return new self($metric, $comparator, $threshold);
    }

    /**
     * @return array<string, mixed>
     */
    public function jsonSerialize(): array
    {
        return $this->toArray();
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        return [
            'metric' => $this->metric,
            'comparator' => $this->comparator,
            'threshold' => $this->threshold,
        ];
    }
}

File: mockserver/mockserver-core/src/main/java/org/mockserver/load/LoadThreshold.java
package org.mockserver.load;

import org.mockserver.model.ObjectWithJsonToString;
import org.mockserver.slo.SloObjective;

/**
 * One in-run pass/fail threshold for a {@link LoadScenario}: a load metric compared against a
 * threshold value. All a scenario's thresholds must hold for the run verdict to be {@code PASS}
 * (logical AND); any breach makes the verdict {@code FAIL}.
 *
 * <p>Thresholds are evaluated on the orchestrator's control tick from <em>per-run</em> data (this
 * run's HDR latency histogram and its request counters) — never from the global
 * {@link org.mockserver.slo.SloSampleStore}, which aggregates all forward traffic rather than this
 * one run. The comparison reuses {@link SloObjective.Comparator} (and its
 * {@link SloObjective#satisfiedBy(double)} semantics) so the two features share one comparator
 * vocabulary, but a load threshold is otherwise independent of the SLO verify feature.
 *
 * @see LoadScenario#getThresholds()
 */
public class LoadThreshold extends ObjectWithJsonToString {

    /**
     * The per-run metric a threshold is evaluated against.
     *
     * <ul>
     *   <li>{@link #LATENCY_P50}/{@link #LATENCY_P95}/{@link #LATENCY_P99}/{@link #LATENCY_P999} —
     *       coordinated-omission-corrected latency percentiles in <b>milliseconds</b>, read from the
     *       run's HDR histogram.</li>
     *   <li>{@link #ERROR_RATE} — failed requests as a <b>fraction</b> of <em>completed</em> requests
     *       (succeeded + failed), in (0..1). In-flight (dispatched-but-not-yet-completed) requests are
     *       excluded so the rate is never diluted by requests whose outcome is not yet known.</li>
     *   <li>{@link #THROUGHPUT_RPS} — requests sent per second over the run's elapsed time.</li>
     * </ul>
     */
    public enum Metric {
        LATENCY_P50,
        LATENCY_P95,
        LATENCY_P99,
        LATENCY_P999,
        ERROR_RATE,
        THROUGHPUT_RPS
    }

    private Metric metric;
    private SloObjective.Comparator comparator;
    private double threshold;

    public static LoadThreshold loadThreshold() {
        return new LoadThreshold();
    }

    public Metric getMetric() {
        return metric;
    }

    public LoadThreshold withMetric(Metric metric) {
        this.metric = metric;
        return this;
    }

    public SloObjective.Comparator getComparator() {
        return comparator;
    }

    public LoadThreshold withComparator(SloObjective.Comparator comparator) {
        this.comparator = comparator;
        return this;
    }

    public double getThreshold() {
        return threshold;
    }

    public LoadThreshold withThreshold(double threshold) {
        this.threshold = threshold;
        return this;
    }

    /**
     * @return true when {@code observed} satisfies this threshold's {@link #comparator} against its
     * {@link #threshold} value, using the same comparison semantics as
     * {@link SloObjective#satisfiedBy(double)}.
     */
    public boolean satisfiedBy(double observed) {
        if (comparator == null) {
            return true;
        }
        return comparator.test(observed, threshold);
    }
}
